### PR TITLE
De-legacify

### PR DESCRIPTION
**Pull request by MangoIV**
*proposal*: move to flakes entirely
- pin all inputs as flake inputs
- remove all of the niv code to remove duplication 
- **add flake-compat code to make those changes compatible with legacy** (nix-shell commands will still work)

### CODE REVIEW

**Comment by MangoIV** [comment]
I see the worries with moving to nix flakes permanently because they're still a new feature, but they have become the quasi standard in packaging software with nix, this is also shown by the statistics. Additionally, this *should* not break any user's experience with the actual legacy nix.

File: README.md [hunk 1]
```diff
@@ -121,6 +125,22 @@ $ niv update nixpkgs -b nixos-unstable
 
 After a brief wait time, the revision is updated.
 
+## Flake support
+
+`ghc.nix` now also has basic flake support, `nixpkgs` and `nixpkgs-unstable` are pinned in the flake inputs, 
+the rest is still managed by `niv` for backwards compatibility. To format all nix code in this repo, run
```
[the remark this answers is not in the document]
Comment (int-index) [reply README.md:131]: What is still managed by `niv`?

**Comment by MangoIV** [comment]
@int-index so sorry, I did not update the README.md in this PR yet, the change is from the other PR, which actually still relies on niv.

**Comment by MangoIV** [comment]
I've done the proper adjustments and added the flake stuff to the ci, let's hope that I get enough ci minutes so it doesn't time out, apart from that I think this should be ready to use.

**Comment by MangoIV** [comment]
the ci is still failing due to some weird issue with running the commands directly and not after entering the shell

**Comment by supersven** [comment]
Hey @MangoIV ,

Thanks a lot for your efforts!

Would you mind to rebase on current `master` branch? I've added a step to the CI that builds an entire GHC and runs some tests.

And, if https://github.com/alpmestan/ghc.nix/pull/130 is now obsolete (?), please consider to close it.

**Comment by MangoIV** [comment]
> And, if https://github.com/alpmestan/ghc.nix/pull/130 is now obsolete (?), please consider to close it.

@supersven I wouldn't consider it obsolete, it's rather an orthogonal proposal  because it's not as intrusive as this PR; it leaves the whole niv workflow intact, whereas with this PR, everything is pinned in the top level flake and the `default.nix` and `shell.nix` are merely there for compatibility. 

I hope that makes sense.

**Comment by MangoIV** [comment]
Build succeeded, hls works, readme is updated, I think this is ready :)

**Comment by sgraf812** [comment]
I've been using this branch for a while now. I think it's safe to merge if the documentation changes are acceptable.

**Comment by supersven** [comment]
> I've been using this branch for a while now. I think it's safe to merge if the documentation changes are acceptable.

Hey @sgraf812 , Thanks a lot for this feedback. It's very helpful. I think I'll give this branch a shot during the next week and then - if it works for me - we can merge it.

**Comment by supersven** [comment]
Hey @MangoIV :wave: 

I've used this branch for a couple of days and can tell that it works. :+1: 

Would you mind to rebase this branch? (There seem to be some conflicts with `master`...)

**Comment by MangoIV** [comment]
Do you want a rebase? I'm getting a ton of conflicts, would be nice to just merge but I understand if that goes against the preferred workflow, then I will take the time.

**Comment by sgraf812** [comment]
I like flakes because it means I can kill `lorri` from my workflow. `use flake ghc.nix` is quite handy.

**Comment by MangoIV** [comment]
Updating and upgrading will be way easier and for the end user flakes just offer the cleaner and more obvious API, at least in my opinion, secondly flakes have already become a quasi standard for packaging nixified repositories; I think you will agree that the main reason why flakes aren't a stabilised feature is because there's not enough funding for nix itself; in terms of adoption they're basically the go-to. 

There's also no downside in using flakes here; in more detail: things like "follows" which are obviously horribly broken aren't used here which would've justified for waiting for them to become a stable feature. 

The only thing I'm worried about is "flake = false" being removed from a future nix version but I'd also count that in the category "less probable".

**Comment by supersven** [comment]
> Do you want a rebase? I'm getting a ton of conflicts, would be nice to just merge but I understand if that goes against the preferred workflow, then I will take the time.

@MangoIV is it still that hard when you squash your commits first (which might be a good idea anyways)?

**Comment by MangoIV** [comment]
Should be done, I hope everything is still as expected. Thank you.

**Comment by MangoIV** [comment]
there you go, squashed and rebased :)

**Comment by MangoIV** [comment]
updating now should merely be `nix flake update`

**Comment by MangoIV** [comment]
https://github.com/DeterminateSystems/update-flake-lock

this might be interesting now.

**Comment by supersven** [comment]
> https://github.com/DeterminateSystems/update-flake-lock
> 
> this might be interesting now. (if you wish I can implement it)

Hm, I'm wondering if it's not just fine to push the `flake.lock` after every manual update? :thinking: 

If you like to investigate a bit further into the GitHub actions CI pipeline: I think one could divide the pipeline into two. One for `nix-shell` and one for `nix develop`. It looks like the `nix-develop` steps just run over the previous results of the `nix-shell` steps. This can lead to just picking up those prior results by GHC. IMHO it would be better if the `nix develop` steps would start from scratch...

**Comment by MangoIV** [comment]
> It looks like the nix-develop steps just run over the previous results of the nix-shell steps. This can lead to just picking up those prior results by GHC. IMHO it would be better if the nix develop steps would start from scratch...

mmmh, I thought about this too, but I didn't see real advantages as if nix holds what it promises it would fail no matter what you executed before; of course this also relies on the GHC being deterministic but that's not what's being tested. the tradeoff would be that the pipeline would take almost twice as long to run. 

if you want, I can pick this up, though.

**Comment by MangoIV** [comment]
> Hm, I'm wondering if it's not just fine to push the flake.lock after every manual update? thinking

it sure is, it would just be nice because nobody had to update manually every time. And we would have everything nice and cached.

**Comment by supersven** [comment]
> mmmh, I thought about this too, but I didn't see real advantages as if nix holds what it promises it would fail no matter what you executed before; of course this also relies on the GHC being deterministic but that's not what's being tested. the tradeoff would be that the pipeline would take almost twice as long to run.

@MangoIV , CI run times are IMHO not a big issue in this project. Most of the time we're only updating dependencies... :smile_cat: (I.e. this is not really about *continuous integration*.)

What the pipeline should show is that the nix env can be used to build a working GHC. I think it would be better to do a fresh build for `nix develop`, because otherwise the steps will likely just pick up what has already been built in the `nix-shell` environment. (*Hadrian*, the GHC build tool knows nothing about Nix... :wink: )

> if you want, I can pick this up, though.

This would be highly appreciated. :+1:

**Comment by supersven** [comment]
This line shows how the binary GHC built in the `nix-shell` env is picked up in the `nix develop` step: https://github.com/alpmestan/ghc.nix/pull/132/checks#step:15:22

**Comment by MangoIV** [comment]
> Hadrian, the GHC build tool knows nothing about Nix... 

that's what I tried to say with "relies on the GHC being deterministic", what I meant was "relies on the GHC being *built* deterministically"

okay, I guess I'll do that then. Maybe not in the next days though :D

**Comment by supersven** [comment]
> > Hm, I'm wondering if it's not just fine to push the flake.lock after every manual update? thinking
> 
> it sure is, it would just be nice because nobody had to update manually every time. And we would have everything nice and cached.

Well, if you like to get into it. We may try it :smiley:

**Comment by bgamari** [comment]
@MangoIV, I'm a bit worried about this direction of travel. Previously, `default.nix` had a number of configuration arguments which one could easily set on the command-line. Doing the same in the post-flakes world would require editing `flake.nix`, which is problematic as I generally have many different trees with different configuration needs (specifically, `nixpkgs`, `bootghc`, and `useClang`). Giving this up seems like an unfortunate loss.

**Comment by MangoIV** [comment]
Hi, I'll be reworking this if you're unhappy with API changes I didn't consider. Sorry for inconveniences caused. Perhaps I might not be able to act on it this or next week though as I currently am under some stress.

**Comment by MangoIV** [comment]
> You could have a default.nix that either uses flake-compat or builtins.getFlake to load the flake, and default.nix would have the same old function arguments, which can be accepted by the legacy nix-shell command.

That is indeed the case. Perhaps something doesn't work as expected anyway?

**Comment by MangoIV** [comment]
Post flakes, it might even be more convenient for you, @bgamari, to define different outputs based on what config option sets you use, I guess that's not as adhoc but certainly easier to reproduce (and self documenting). If that's not convenient for you though, I will definitely have to fix it, obviously.

**Comment by Ericson2314** [comment]
> I can simply refrain from pulling until things have been sorted out.

That's what I did, lol.

**Comment by MangoIV** [comment]
So, I have done a bit of research, as far as I can tell, it's not visible in the near future to make passing arguments to nix flakes possible; the main counter argument is that it will hurt provenance to allow so; every time you pass an argument, you potentially dirty the lock file. (I personally do not think this is a valid argument because as of today we already have mechanisms that prevent executing commands that would dirty the lockfile in a context where this is not explicitly asked for)

I will ask around a bit more and will look at the implementation of `flake-compat` so we can at least get back passing arguments for the legacy nix commands. 

I will also try to think of alternative solutions such as providing a build matrix that combines some of the most common arguments or adding an input from a flake that has some possible configuations as outputs and can just be overridden in an adhoc manner (I think this is not really feasible as it's probably too complex)

**Comment by MangoIV** [comment]
I think I have a solution to the regression wrt. legacy commands, I will open a PR after documenting it properly (it's a bit hacky, not much, but a bit) 

Sorry that it took so long @bgamari.

**Pull request merged**